### PR TITLE
boot: do not reorder boot assets when generating predictable boot chains and other small tweaks

### PR DESCRIPTION
**Pull request by bboozzoo**
Boot assets should not be reordered when predictable boot chains are generated.
However, the list of hashes inside in a boot asset, needs to be sorted.

### CODE REVIEW

File: boot/bootchain_test.go [hunk 1]
```diff
@@ -302,25 +228,52 @@ func (s *bootchainSuite) TestBootChainEqualForResealComplex(c *C) {
...
+
+	// already sorted
 	pbOther := boot.PredictableBootChains{
 		{
 			BrandID:        "mybrand",
 			Model:          "foo",
 			Grade:          "dangerous",
 			ModelSignKeyID: "my-key-id",
 			AssetChain: []boot.BootAsset{
+				{Role: bootloader.RoleRunMode, Name: "loader", Hashes: []string{"c", "d"}},
```
Comment (pedronis) [inline boot/bootchain_test.go:243]: can we make some, all of these tests realistic,  RunMode would come generally after recovery?

File: boot/bootchain_test.go [hunk 2]
```diff
@@ -239,12 +167,12 @@ func (s *bootchainSuite) TestBootChainMarshalFull(c *C) {
 		Model:          "foo",
 		Grade:          "dangerous",
 		ModelSignKeyID: "my-key-id",
-		// assets are sorted
+		// assets are not reordered
 		AssetChain: []boot.BootAsset{
-			{Role: bootloader.RoleRecovery, Name: "loader", Hashes: []string{"d"}},
+			{Role: bootloader.RoleRunMode, Name: "loader", Hashes: []string{"c", "d"}},
 			// hash lists are sorted
```
Comment (mvo5) [inline boot/bootchain_test.go:173]: (nitpick²) you could move the comment now one line up I think

**Review by mvo5 — APPROVED** [review]
LGTM

File: boot/bootchain_test.go [hunk 3]
```diff
@@ -302,25 +228,52 @@ func (s *bootchainSuite) TestBootChainEqualForResealComplex(c *C) {
...
+	eq = boot.PredictableBootChainsEqualForReseal(pb, pbOther)
+	c.Check(eq, Equals, true, Commentf("not equal\none: %v\nother: %v", pb, pbOther))
+
+	pbNotEqual := boot.PredictableBootChains{
+		{
+			BrandID:        "mybrand",
+			Model:          "foo",
+			Grade:          "dangerous",
+			ModelSignKeyID: "my-key-id",
+			// order of boot assets is different
```
Comment (pedronis) [inline boot/bootchain_test.go:261]: I don't understand the comment here

File: boot/bootchain_test.go [hunk 4]
```diff
@@ -302,25 +228,52 @@ func (s *bootchainSuite) TestBootChainEqualForResealComplex(c *C) {
...
+			AssetChain: []boot.BootAsset{
 				{Role: bootloader.RoleRecovery, Name: "shim", Hashes: []string{"a", "b"}},
+
```
Comment (pedronis) [inline boot/bootchain_test.go:264]: why the newline?

**Review by pedronis — COMMENTED** [review]
I think the test unless they are checking something in particular should use realistc order for bootAssets
with recovery roles coming before run-mode roles

File: boot/bootchain_test.go (hunk 3 above)
Comment (bboozzoo) [reply boot/bootchain_test.go:261]: Clarified in the code

**Review by pedronis — APPROVED** [review]
thank you, pushed some more tweaks myself

**Review by mvo5 — APPROVED** [review]
Still looks good

**Pull request merged**